src/ophyd_async/tango/base_devices/_base_device.py

```diff
@@ -150,76 +93,62 @@ def decorator(cls):
-            type_args = get_args(device_type)
-            datatype = type_args[0] if type_args else None
-            backend = make_backend(datatype=datatype, device_proxy=device.proxy)
-            setattr(device, name, origin(name=name, backend=backend))
-
-        elif issubclass(origin, Device) or isinstance(origin, Device):
-            assert callable(origin), f"{origin} is not callable."
-            setattr(device, name, origin())
-
-
-async def _fill_proxy_entries(device: TangoDevice):
-    if device.proxy is None:
-        raise RuntimeError(f"Device proxy is not connected for {device.name}")
-    proxy_trl = device.trl
-    children = [name.lstrip("_") for name, _ in device.children()]
-    proxy_attributes = list(device.proxy.get_attribute_list())
-    proxy_commands = list(device.proxy.get_command_list())
-    combined = proxy_attributes + proxy_commands
-
-    for name in combined:
-        if name not in children:
-            full_trl = f"{proxy_trl}/{name}"
-            try:
-                auto_signal = await __tango_signal_auto(
-                    trl=full_trl, device_proxy=device.proxy
+class TangoDeviceBackend(DeviceBackend):
+    def __init__(
+        self,
+        device_type: type[Device],
+        trl: str | None,
+        device_proxy: DeviceProxy | None,
+        polling: tuple[bool, float, float | None, float | None],
+        signal_polling: dict[str, tuple[bool, float, float, float]],
+    ) -> None:
+        super().__init__(device_type)
+        self.trl = trl
+        self.proxy = device_proxy
+        self._polling = polling
+        self._signal_polling = signal_polling
+        self._filler = DeviceFiller(
+            children=self.children,
+            device_type=device_type,
+            signal_backend_type=TangoSignalBackend,
+            device_backend_type=TangoDeviceBackend,
+        )
+
+    async def connect(self, mock: bool, timeout: float, force_reconnect: bool) -> None:
+        if mock:
+            # Make 2 entries for each DeviceVector
+            self._filler.make_soft_device_vector_entries(2)
+        else:
+            if self.trl and self.proxy is None:
+                self.proxy = await AsyncDeviceProxy(self.trl)
+            elif self.proxy and not self.trl:
+                self.trl = self.proxy.name()
+            else:
+                raise TypeError("Neither proxy nor trl supplied")
+
+            children = sorted(
+                set()
+                .union(self.proxy.get_attribute_list())
+                .union(self.proxy.get_command_list())
+            )
+            for name in children:
+                # TODO: strip attribute name
+                full_trl = f"{self.trl}/{name}"
+                signal_type = await infer_signal_type(full_trl, self.proxy)
+                if signal_type:
+                    backend = self._filler.make_child_signal(name, signal_type)
+                    backend.datatype = await infer_python_type(full_trl, self.proxy)
```

Comment:
@burkeds if I remove the `infer_python_type` it still works with `TangoSignalBackend.datatype = None`. Is there any reason to keep `infer_python_type`?

Reply:
They will initialize but the datatype of the signal will be None. This should throw up an error if you try to say, write a float to the signal. infer_python_type makes sure automatically generated signals are typed according to the signal's Tango type.

Reply:
What should throw up an error? Signal or the Tango proxy?

Reply:
I would say that Signal should throw up the error. The proxy knows what type it should be, so it's the signal configuration which is the issue. I think that currently, the type mismatch is checked in `get_trl_descriptor`.

Reply:
I think we need a followup PR to look at the Tango tests, let's get this one merged first though